File: src/main/java/ThirtyDayLeetCodingChallenge/Day13_ContiguousArray.java
package ThirtyDayLeetCodingChallenge;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Day13_ContiguousArray {
    public static void main(String[] args) {
        System.out.println(findMaxLength(new int[]{0, 1}));
        System.out.println(findMaxLength(new int[]{0, 1, 0}));
        System.out.println(findMaxLength(new int[]{0, 1, 0, 0, 0, 0, 1, 0, 0, 1}));
    }

    public static int findMaxLength(int[] nums) {
        int zeros = (int) Arrays.stream(nums).filter(number -> number == 0).count();
        int ones = nums.length - zeros;

        if (zeros == ones) return nums.length;
        else if (zeros == nums.length || ones == nums.length) return 0;
        else {
            zeros = 0;
            ones = 0;
            int max = 0;
            Map<Integer, Integer> memo = new HashMap<>();
            memo.put(0, -1);

            for (int i = 0; i < nums.length; i++) {
                if (nums[i] == 0) zeros++;
                else ones++;

                if (memo.containsKey(zeros - ones)) max = Math.max(max, i - memo.get(zeros - ones));
                else memo.put(zeros - ones, i);
            }

            return max;
        }
    }
}

File: src/main/kotlin/ThirtyDayLeetCodingChallenge/Day13_ContiguousArray.kt
import java.util.*

fun main() {
    println(findMaxLength(intArrayOf(0, 1)))
    println(findMaxLength(intArrayOf(0, 1, 0)))
    println(findMaxLength(intArrayOf(0, 1, 0, 0, 0, 0, 1, 0, 0, 1)))
}

fun findMaxLength(nums: IntArray): Int {
    var zeros = Arrays.stream(nums).filter { number: Int -> number == 0 }.count().toInt()
    var ones = nums.size - zeros

    return if (zeros == ones) nums.size else if (zeros == nums.size || ones == nums.size) 0 else {
        zeros = 0
        ones = 0
        var max = 0
        val memo: MutableMap<Int, Int> = HashMap()
        memo[0] = -1
        for (i in nums.indices) {
            if (nums[i] == 0) zeros++
            else ones++

            if (memo.containsKey(zeros - ones)) max = Math.max(max, i - memo[zeros - ones]!!)
            else memo[zeros - ones] = i
        }
        max
    }
}
